src/session-pool.js

```diff
@@ -435,252 +366,370 @@ SessionPool.prototype.requestStream = function(config) {
-      var err = new Error('Unable to acquire Session, timeout occurred.');
-      var acquire;
-
-      for (var i = self.pendingAcquires.length - 1; i > -1; i--) {
-        acquire = self.pendingAcquires[i];
-        acquire.timeout -= intervalSpeed;
-
-        if (acquire.timeout < 0) {
-          self.pendingAcquires.splice(i, 1);
-          acquire.callback(err);
-        }
-      }
-    }
+  function killIntervals() {
+    clearInterval(self.pingHandle_);
+    clearInterval(self.evictHandle_);
   }
 };
 
 /**
- * Release a write session. This will also begin a new transaction before
- * returning it to the pool.
+ * Makes a keep alive request to all the idle sessions.
  *
  * @private
  *
- * @param {Session} session - The session to be released.
+ * @fires SessionPool#error
  * @returns {Promise}
  */
-SessionPool.prototype.releaseWriteSession_ = function(session) {
+SessionPool.prototype.pingIdleSessions_ = function() {
   var self = this;
-  var transaction = self.createTransaction_(session);
 
-  return transaction
-    .begin()
-    .then(function() {
-      session.transaction_ = transaction;
-      return self.writePool.release(session);
-    })
-    .then(null, function() {
-      return self.writePool.destroy(session);
-    });
+  var sessions = this.getIdleSessions_();
+  var requests = sessions.map(function(session) {
+    return self.requestQueue_
+      .add(function() {
+        return session.keepAlive();
+      })
+      .catch(function(err) {
+        self.destroySession_(session);
+        self.emit('error', err);
+      });
+  });
+
+  return Promise.all(requests);
 };
 
 /**
- * Check if a session is dead.
+ * Prepares a transaction.
  *
  * @private
  *
- * @class
- *
- * @param {Pool} pool genericPool instance.
+ * @param {Session} session The session to prepare the transaction for.
+ * @param {TransactionOptions} [options] The transaction options.
+ * @returns {Promise<Transaction>}
  */
-function SessionEvictor(pool) {
-  this.pool = pool;
-  this.evictor = new genericPool.DefaultEvictor();
-}
+SessionPool.prototype.prepareTransaction_ = function(session, options) {
+  var self = this;
 
-/**
- * Check if the ping interval has been met. If so, attempt to run a dummy query.
- * If the query fails, we'll consider the session dead.
- *
- * @return {boolean}
- */
-SessionEvictor.prototype.evict = function(config, resource, available) {
-  var session = resource.obj;
+  var transaction = session.transaction(options);
+  var end = transaction.end;
 
-  if (session.evicted_) {
-    return true;
-  }
+  transaction.end = function(callback) {
+    self.release(session);
+    return end.call(transaction, callback);
+  };
 
-  var needsKeepAlive = this.evictor.evict(config, resource, available);
+  return transaction.begin().then(
+    function() {
+      session.txn = transaction;
+      return transaction;
+    },
+    function(err) {
+      transaction.end();
+      self.destroySession_(session);
```

Comment:
only destroy if the error is "NOT_FOUND", otherwise put the session back in the read only state.